src/txn/txn.c

```diff
@@ -208,8 +211,15 @@ __txn_get_snapshot_int(WT_SESSION_IMPL *session, bool publish)
 
     /* Fast path if we already have the current snapshot. */
     if ((commit_gen = __wt_session_gen(session, WT_GEN_COMMIT)) != 0) {
-        if (F_ISSET(txn, WT_TXN_HAS_SNAPSHOT) && commit_gen == __wt_gen(session, WT_GEN_COMMIT))
+        WT_ASSERT(session, F_ISSET(txn, WT_TXN_HAS_SNAPSHOT));
+        if (!F_ISSET(txn, WT_TXN_REFRESH_SNAPSHOT) &&
+          commit_gen == __wt_gen(session, WT_GEN_COMMIT))
             return;
+        /*
+         * Leave the generation here and enter again later to acquire a new snapshot if any
+         * concurrent transactions after we entered the generation have been committed and changed
+         * the global commit generation.
+         */
         __wt_session_gen_leave(session, WT_GEN_COMMIT);
```

Comment:
I suspect it isn't safe to leave the generation even temporarily, while acquiring a refreshed snapshot. Though I'd be happy to be convinced otherwise.

Reply:
I do not see any issue here to leave the generation and enter again.

Reply:
@agorrod My plan is to refactor the code around snapshot( `get_snapshot / release_snapshot` ) to use `WT_GEN_HAS_SNAPSHOT ` generation  to enter/leave.   

Introduce a new snapshot (for ex - `WT_GEN_COMMIT ` ) for [WT-9199](https://jira.mongodb.org/browse/WT-9199).
Please let me know if you have suggestion.